### PR TITLE
Add more test cases for R Studio notebook 

### PR DESCRIPTION
**Pull request by atheo89**
## Description
This is an enhancement on the test framework for RStudio notebook. 

The new tests include:
- Create, save, and execute R script.
- Evaluate if can be download R package 

## How Has This Been Tested?

1. Copy your login cmd from your ocp cluster to your terminal
2. Run `make rstudio-c9s-python-3.9  -e IMAGE_REGISTRY=quay.io/${YOUR_QUAY_USER}/workbench-images`
3. Run `make deploy-c9s-rstudio-c9s-python-3.9`
4. Give a bit of time till the pod be in running status and run `make validate-rstudio-image image=rstudio-c9s-python-3.9-2023b_20230129` (check to be the correct tag)
5. Once you evaluate/enhance the testing run `make undeploy-c9s-rstudio-c9s-python-3.9` to destroy the testing pod

Bellow are are the logs from the `make validate-rstudio-image image=rstudio-c9s-python-3.9-2023b_20230129`  execution:
```
# Running tests for  RStudio image...
bin/kubectl wait --for=condition=ready pod rstudio-pod --timeout=300s
pod/rstudio-pod condition met
=> Checking container image rstudio-c9s-python-3.9-2023b_20230129 for package intallation...
Tinytex installation successful!
=> Checking container image rstudio-c9s-python-3.9-2023b_20230129 for curl...
curl executed successfuly!
=> Checking container image rstudio-c9s-python-3.9-2023b_20230129 for python...
python executed successfuly!
=> Checking container image rstudio-c9s-python-3.9-2023b_20230129 for oc...
oc executed successfuly!
=> Checking container image rstudio-c9s-python-3.9-2023b_20230129 for /usr/lib/rstudio-server/bin/rserver...
/usr/lib/rstudio-server/bin/rserver executed successfuly!
=> Creating R file and executing on the container...
R script executed successfully!
```
## Merge criteria:

- [ ] The commits are squashed in a cohesive manner and have meaningful messages.
- [ ] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [ ] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

File: Makefile
Comment (jiridanek) [inline Makefile:475]: The indentation is strange. Shouldn't the else be aligned under the if to which it belongs?

File: Makefile
Comment (jiridanek) [inline Makefile:472]: Is it enough to install the package? To check whether the installation was successful, I think it is also necessary to at least try to `include()` the package afterwards from some script. `R -e "include(..."`

File: Makefile
Comment (atheo89) [reply Makefile:475]: It is shown like this only on git preview, if you clone the PR it is fine, could you evaluate from you side as well?

File: Makefile
Comment (jiridanek) [reply Makefile:472]: The reason there is `/opt/app-root/src/R/temp-library` is that without it, the installation would fail because user does not have write access?

Comment (jiridanek) [inline Makefile:472]: The reason there is /opt/app-root/src/R/temp-library is that without it, the installation would fail because user does not have write access?

Comment (atheo89) [reply Makefile:472]: Exactly, because we change the permissions on this line -> https://github.com/atheo89/notebooks/blob/main/rstudio/c9s-python-3.9/Dockerfile#L113

Comment (atheo89) [reply Makefile:472]: Yes this could be good addition, for now we are checking is we have exit code ==1

File: Makefile
Comment (jiridanek) [reply Makefile:475]: https://patch-diff.githubusercontent.com/raw/opendatahub-io/notebooks/pull/392.diff looks wrong too,

image

ok, I'll go clone this and hopefully it will align itself better then

File: Makefile
Comment (jiridanek) [reply Makefile:472]: So if I want to include  what I've installed in my script, I need to keep specifying this path in every include()? Doesn't R have virtualenvs? Or is this how venvs in R work?

File: Makefile
Comment (atheo89) [reply Makefile:475]: Let us know, this how it looks on my vs code
image

File: Makefile
Comment (atheo89) [reply Makefile:472]: This temporary lib is only for the testing on the container. When I was doing manual testing on RStudio IDE itself i didn't have to create particular env for package installations.

File: Makefile
Comment (jiridanek) [reply Makefile:475]: I see the same, but my background color is much nicer than yours ;ddd

image

What #!^% is going on with the GitHub previews?!

Comment (jiridanek) [reply Makefile:475]: Got it, mixing tabs and spaces, and configured tab size!

image

Is it possible to be more consistent regarding tabs/spaces?

Comment (atheo89) [reply Makefile:475]: The mystery solved!

File: Makefile
Comment (jiridanek) [reply Makefile:472]: >  When I was doing manual testing on RStudio IDE

Why, what is different from the situation now?

Comment (atheo89) [reply Makefile:472]: not sure

Comment (jiridanek) [reply Makefile:472]: 1? I'd expect successful return code to be 0, https://tldp.org/LDP/abs/html/exitcodes.html

Comment (atheo89) [reply Makefile:472]: Yeah, exit code 1 to catch the error

**Comment by jiridanek** [comment]
General workflow suggestion: What's the point of having a `How Has This Been Tested?` docs on test PRs?

Since we care whether the tests run in CI or not, I think the most practical way to test tests is to create a PR, wait for CI to run the added tests, and then add a link to the test report from the CI.

The instructions on how to run tests locally on your own machine should be in a .md file in the repository, not repeated every time on every PR.

File: Makefile
Comment (jiridanek) [reply Makefile:475]: nice, thanks!

**Comment by atheo89** [comment]
> General workflow suggestion: What's the point of having a `How Has This Been Tested?` docs on test PRs?
> 
> Since we care whether the tests run in CI or not, I think the most practical way to test tests is to create a PR, wait for CI to run the added tests, and then add a link to the test report from the CI.
> 
> The instructions on how to run tests locally on your own machine should be in a .md file in the repository, not repeated every time on every PR.

We need enhancements for sure! However, this section serves as a mandatory requirement for the developer to document the testing process steps. It is crucial to provide clear instructions to the reviewer, ensuring they can easily catch up and replicate the pull request on their end.

File: Makefile
Comment (jiridanek) [reply Makefile:472]: anyways, the test code looks good

```
	if [ $$? -eq 0 ]; then \
            echo "Tinytex installation successful!"; \
```

> Yes this could be good addition

I've added it to the tasks list on https://issues.redhat.com/browse/RHOAIENG-2308 for now. We can split it off to another task if that's what others want.

**Comment by jiridanek** [comment]
> We need enhancements for sure! However, this section serves as a mandatory requirement for the developer to document the testing process steps.

Sure, but since what you are adding here are automated test scripts, it seems unproductive to me to repeatedly document the steps needed for their execution in every PR raised. This should be done once, in a https://github.com/opendatahub-io/notebooks/blob/main/docs/developer-guide.md or a separate document linked to it, and then it can be referenced as needed.

Besides, what is important is not checking that the tests run as expected on people's laptops, but that the tests run on the CI. Reviewers should not need to manually repeat automated tests that the CI is already running. (CI is running this, I hope).

File: Makefile
Comment (harshad16) [inline Makefile:456]: It would be best if we bring this out of Makefile
and move it to `test_something.R` file

as it would easy to maintain that , then R code in Makefile.

**Review by harshad16 — COMMENTED** [review]
Great work,
one small comment

File: Makefile
Comment (atheo89) [reply Makefile:456]: Agree, we may use `curl -sSL -o test.R "https://raw.githubusercontent.com/.../test.R"` 
Where we can host the test.R on `/rstudio/c9s-python-3.9/test/test.R` as we do for the Jupyter notebooks

File: Makefile
Comment (jiridanek) [inline Makefile:498]: Another misaligned if/then/else

File: Makefile
Comment (jiridanek) [inline Makefile:455]: And one more misaligned if/then/else. Here, writing it in separate file should help.

(I guess VS Code puts in the tabs because Makefile requires tabs in some places. And the Makefile support in VS Code does not work well with these embedded scripts.)

File: Makefile
Comment (harshad16) [inline Makefile:461]:
```suggestion
	curl -sSL -o test_script.R "${NOTEBOOK_REPO_BRANCH_BASE}/rstudio/c9s-python-3.9/test/test_script.R" > /dev/null 2>&1 ; \
```

**Review by jiridanek — APPROVED** [review]

**Review by harshad16 — APPROVED** [review]
/lgtm
/approve

thanks :100:

**Pull request merged**
